### ISSUE TITLE
Option to send Ctrl+Alt+Del or Ctrl+Alt_End for nested RDP sessions

### ISSUE THREAD

**Issue by yderi** [mRemoteNG/mRemoteNG#1349]
Hello everyone.
I have enabled the "Key combination" option and it works fine to me. But when you open a secondary RDP session to go further (open RDP connection again on remote side) it cannot pass Ctrl+Alt+End combination for the second or third nested session.
It works only for the first one. Some time we have even three or four nested RDP session, it such a kind of security using jump-boxes to go thru.
The trick here is to open on-screen keyboard in the first session and using it, send a Ctrl+Alt+End combination to the secondary nested session (and so on).
It would be very nice to have a built-in option to send such a predefined key combination for a nested session. (For example by right clicking on connection tab and call this function)
Thanks a lot

**Comment by another-git-hub-user** [comment]
I do not believe this is something that mRemoteNG can do, nor any RDP client. Have you considered using an RD Gateway to connect through instead of using the nested approach? I would expect a mammoth improvement in performance and responsiveness of the RDP sessions with a gateway.

Another option is to set up custom keyboard shortcuts within the descendants that would provide the equivalent of Ctrl+Alt+Del. Perhaps Autohotkey (https://www.autohotkey.com) which makes it trivial to create shortcuts would be an option on each remote machine.

**Comment by nmat** [comment]
This is hard to fix I think due to how RDP is built up.

I do not even think that a rdp -> rdp -> rdp would be able to do that to understand which connection is the one that is active to recieve the keyboard shortcuts.

**Comment by yderi** [comment]
Hello
Remote Desktop Connection Manager can send a such key combination
(For example: RDCMan v2.7 build 1406.0)
![SendKey_Options](https://user-images.githubusercontent.com/38353406/55144388-69bd8000-5149-11e9-88b7-94e46f533fed.jpg)

I think you misunderstood this function, we should not send this key combination directly from a keyboard thru all nested RDP connections. Just use it on demand (in appropriate nested level, no matter how it is deep) by clicking on the tab connection or from the appropriate menu. For Example like RDC Manager do this.

**Comment by sparerd** [comment]
For implementation: https://docs.microsoft.com/en-us/windows/desktop/TermServ/imsrdpclientnonscriptable-sendkeys

**Comment by TheDevilDan** [comment]
Work Around : launch "osk.exe" (virtual keyboard), do CTRL + ALT with the keybord, and "SUPPR" with the virtual keyboard

**Comment by Elddric** [comment]
as TheDevilDan says you can use this workaround: launch the virtual keyboard with osk.exe or click start >windows ease of access > Onscreen keyboard. On your physical keyboard do CTRL+ALT . Then on your Virtual keyboard press Delete. This worked for me.

**Comment by bc185161** [comment]
In Config for connection, under Redirect, set Key Combinations to Yes

answer from here: https://github.com/mRemoteNG/mRemoteNG/issues/231

**Comment by thecosmictrickster** [comment]
> 
> 
> as TheDevilDan says you can use this workaround: launch the virtual keyboard with osk.exe or click start >windows ease of access > Onscreen keyboard. On your physical keyboard do CTRL+ALT . Then on your Virtual keyboard press Delete. This worked for me.

This usually works but not quite as convenient. Another wrinkle is if osk does not start for reasons unknown. I had this recently, and ended up having to copy RDCMan to the jump host in question so I could use the [Send keys] functionality.

**Comment by HelpITru** [comment]
cross link https://github.com/mRemoteNG/mRemoteNG/issues/1455

**Comment by asrdias** [comment]
> Hello everyone. I have enabled the "Key combination" option and it works fine to me. But when you open a secondary RDP session to go further (open RDP connection again on remote side) it cannot pass Ctrl+Alt+End combination for the second or third nested session. It works only for the first one. Some time we have even three or four nested RDP session, it such a kind of security using jump-boxes to go thru. The trick here is to open on-screen keyboard in the first session and using it, send a Ctrl+Alt+End combination to the secondary nested session (and so on). It would be very nice to have a built-in option to send such a predefined key combination for a nested session. (For example by right clicking on connection tab and call this function) Thanks a lot

It can....

Under Redirect > Key Combinations > Set to Yes

Alt + CTRL + END will be ALT + CTRL + DEL

**Comment by mpapini64** [comment]
ngRemote will never make this change they are stuck on version v1.77.3.1784-NB for 2 years now and I think the project is finished, too bad.
Anyway the solution to send Ctrl-Alt-END is the following (OSK does not work)
- open: PowerShell
- type: (New-Object -COM Shell.Application).WindowsSecurity()

**Comment by robertpopa22** [comment]
**Fix available for testing**

A fix for this issue has been implemented in our community fork (`robertpopa22/mRemoteNG`, branch `main`). A beta build including this fix is available from the fork's Releases page.

Please test and report if this resolves your issue.

---

> *We sincerely appreciate the mRemoteNG maintainers and the wonderful community that has kept this project alive and thriving for over a decade. Your work powers thousands of IT professionals worldwide, including us. Our fork builds entirely on your foundation — we're preparing a consolidated Pull Request to contribute these improvements back to the official project. Thank you for everything you've built.*
>
> *— robertpopa22, community contributor*
